### ISSUE TITLE
Feature request: SecurityCookieAccessor async

### ISSUE THREAD

**Issue by bmorris591** [ArcBees/GWTP#587]
Currently the GWTP RPC CSRF protection framework works by (assuming `JSESSIONID` is not used):
1. a filter issues a random cookie with a specified name
2. on the client side this cookie is echoed into the RPC request having retrieved it with the help of the `SecurityCookieAccessor`
3. on the server side this value is compared to the value of the cookie

This has the upside of being simple, but it is somewhat vulnerable to CSRF - see issues [#484](https://github.com/ArcBees/GWTP/issues/484) and [#485](https://github.com/ArcBees/GWTP/issues/485). It also requires the cookie to not be `httpOnly`, which means even with those issues fixed, CSRF is still a possibility - just not session fixation.

GWT already has a built in framework for CSRF protection, which works thus:
1. a special servlet, the `XsrfTokenServiceServlet` is configured on the server side, this servlet's job is to server a NONCE via an RPC request
2. the client first calls the Xsrf servlet to get a NONCE
3. the client then sets that NONCE in the RPC request it wishes to perform
4. the server validates this NONCE

This has the advantage of not requiring a non-`httpOnly` cookie to be set. It also requires that a request is sent to the server, the response read and used to generate a valid RPC request; this precludes CSRF. It is also purely server-side, so it's fairly flexible. Currently this is simply implemented as an MD5 hash of the session id, but the end user can fairly easily change it to single use, expiring, tokens (for example).

Having two separate mechanisms seems a little redundant, it adds configuration burden when using both RPC and GWTP RPC.

Looking through the source code, the only thing preventing integration with the GWT CSRF framework is the fact that `SecurityCookieAccessor` is synchronous.

If the source were changed to make `SecurityCookieAccessor` work asynchronously then two implementations could be provided - one that simply plugged into the GWT CSRF framework and one that used the existing GWTP CSRF framework.

**Comment by bradcupit** [comment]
If you use RandomSecurityCookieFilter instead of HttpSessionSecurityCookieFilter (related to #484) then you can make the JSESSIONID HttpOnly, which means even in an XSS attack you're still not vulnerable to session fixation.

We just need RandomSecurityCookieFilter to be a secure cookie (#485) so it's not vulnerable to sniffing. (In the GWTP project I'm on we have our own filter, similar to RandomSecurityCookieFilter, that generates a secure cookie).

I think this gives all the CSRF protection that the GWT nonce approach does.

In an XSS attack the random security cookie is vulnerable (since it can't be HttpOnly), but if you're already vulnerable to XSS then the attacker can get the nonce from the server.

In short: I *think* we could just fix #485 and be good.

**Comment by bmorris591** [comment]
@bradcupit  you're right, that would solve the immediate problem. My main feature request is for a `SecurityCookieAccessorAsync` so that we can _choose_ what to use. Fixing #485 would secure the GWTP CSRF framework, but I would like to, ideally, only have one CSRF framework in my application.

**Comment by semteXKG** [comment]
Is there anything planned in the (nearer) future concerning this issue? I can't use the RandomSecurityCookieFilter as it apparently [has issues with concurrency (known issue?)](https://groups.google.com/forum/#!topic/gwt-platform/BzaOaepdglU) which I can confirm. The `HttpSessionSecurityCookieFilter` was removed with 1.5.3, leaving us stranded. I'm currently thinking of bringing back something like the `HttpSessionSecurityCookieFilter` but using a hash of the session id instead of the real thing. this won't solve all issues but it's better than nothing.

**Comment by WTBenjamin** [comment]
@semteXKG Did you realize your hash cookie filter? We just did the update to GWT 2.8.2 and GWTP 1.6 and are facing the same concurrency issues you mentioned with the `RandomSecurityCookieFilter`. As we have a pretty big application to maintain we just can not switch to GWT RPC. Only chance for us here is to write an own CookieFilter or revert back to the old versions of GWT and GWTP, since we have very little hope to get an 'official' fix for the concurrency problem.

 We would really appreciate if you (or someone else) could share your filter.
